src/source/test_waveform.rs

```diff
@@ -0,0 +1,185 @@
+//! Generator sources for various periodic test waveforms.
+//!
+//! This module provides several periodic, deterministic waveforms for testing other sources and
+//! for simple additive sound synthesis. Every source is monoaural and in the codomain [-1.0f32,
+//! 1.0f32].
+//!
+//! # Example
+//!
+//! ```
+//! use rodio::source::{TestWaveform,Function};
+//!
+//! let tone = TestWaveform::new(cpal::SampleRate(48000), 440.0, Function::Sine);
+//! ```
+use std::f32::consts::TAU;
+use std::time::Duration;
+
+use super::SeekError;
+use crate::Source;
+
+/// Waveform functions.
+#[derive(Clone, Debug)]
+pub enum Function {
+    /// A sinusoidal waveform.
+    Sine,
+    /// A triangle waveform.
+    Triangle,
+    /// A square wave, rising edge at t=0.
+    Square,
+    /// A rising sawtooth wave.
+    Sawtooth,
+}
+
+impl Function {
+    /// Create a single sample for the given waveform
+    #[inline]
+    fn render(&self, i: u64, period: f32) -> f32 {
+        let cycle_pos: f32 = i as f32 / period;
+
+        match self {
+            Self::Sine => (TAU * cycle_pos).sin(),
+            Self::Triangle => 4.0f32 * (cycle_pos - (cycle_pos + 0.5f32).floor()).abs() - 1f32,
+            Self::Square => {
+                if cycle_pos % 1.0f32 < 0.5f32 {
+                    1.0f32
+                } else {
+                    -1.0f32
+                }
+            }
+            Self::Sawtooth => 2.0f32 * (cycle_pos - (cycle_pos + 0.5f32).floor()),
+        }
+    }
+}
+
+/// An infinite source that produces one of a selection of test waveforms.
+#[derive(Clone, Debug)]
+pub struct TestWaveform {
+    sample_rate: cpal::SampleRate,
+    period: f32,
+    function: Function,
+    i: u64,
+}
+
+impl TestWaveform {
```

Comment:
To me `Test` implies its only for debugging, however I think this is more useful then that. I'am also afraid of ppl skipping it when they see `Test` in the name, assuming its not reliable/performant/stable. We could call it `SimpleWaveform`? How does that sound to you?

Reply:
Other suitable terms might be `SignalGenerator`, `WaveformGenerator`, `FunctionGenerator`, `ToneGenerator`. Maybe ToneGenerator...

Reply:
all very good names, damn.. 

My personal favorite is `SignalGenerator`, I have a background in physics and thats how we called the device that generates sawtooth/sine etc. The choice is yours :+1:

Reply:
`SignalGenerator` is good, we don't reserve the word "signal" for anything special to the framework so yeah that works for me.